src/main/java/org/opensearch/security/privileges/ActionPrivileges.java

```diff
@@ -407,7 +414,56 @@ PrivilegesEvaluatorResponse providesPrivilege(PrivilegesEvaluationContext contex
                 }
             }
 
-            return PrivilegesEvaluatorResponse.insufficient(action);
+            // 4: Evaluate api tokens
+            return apiTokenProvidesClusterPrivilege(context, Set.of(action), false);
+        }
+
+        /**
+         * Evaluates cluster privileges for api tokens. It does so by checking exact match, regex match, * match, and action group match in a non-optimized, naive way.
+         * First it expands all action groups to get all the actions and patterns of actions. Then it checks * if not an explicit check, then for exact match, then for pattern match.
+         */
+        PrivilegesEvaluatorResponse apiTokenProvidesClusterPrivilege(
+            PrivilegesEvaluationContext context,
+            Set<String> actions,
+            Boolean explicit
+        ) {
+            Permissions permissions = context.getPermissionsForApiToken();
+            Set<String> resolvedClusterPermissions = actionGroups.resolve(permissions.getClusterPerm());
+
+            // Check for wildcard permission
+            if (!explicit) {
+                if (resolvedClusterPermissions.contains("*")) {
+                    return PrivilegesEvaluatorResponse.ok();
+                }
+            }
+
+            // Check for exact match
+            if (!Collections.disjoint(resolvedClusterPermissions, actions)) {
+                return PrivilegesEvaluatorResponse.ok();
+            }
+
+            // Check for pattern matches (like "cluster:*")
+            for (String permission : resolvedClusterPermissions) {
+                // skip pure *, which was evaluated above
+                if (permission != "*") {
```

Comment:
`!=` operator shouldn't be used on strings.

Is this check necessary? When `permission` is `*` it creates a Matcher that always returns true without having to do any regex or glob matching. I think this case should be covered by L454-459

Reply:
This is to cover the case where when we want to check for explicit permission matching, if it was rejected in the first block, we don't want it to pass the wildcard matcher below.

Reply:
Switched to .equals() in: [685dcfc](https://github.com/opensearch-project/security/pull/4992/commits/685dcfc8b7752fc4b118b1b85e3b9e6f827e7c0d), thanks!